### ISSUE TITLE
PNG images will stop being read if the source callback doesn't return the same number of bytes as said to be the maximum buffer size

### ISSUE THREAD

**Issue by vibbix** [libvips/libvips#1889]
**Describe the bug**
When using the streaming API, PNG loader will not tolerate different sized reads from what as requested.

**To Reproduce**
Load a PNG Image from a stream using the source API. If the callbacks  are less than the size required, libspng will return with a "pngload_source: end of stream" error. In older lib VIPS version, this creates a lot of "[WARNING] not enough data"

**Steps to reproduce the behavior**:
1. Load a PNG image using the Source API
2. Have the source only partially fill the buffer from the callback function
Other loaders, like JPEG seem to tolerate this behavior.

**Expected behavior**
The png loader should be able to tolerate getting only partial responses back, it seems to ignore the number of bytes returned back to it.

**Actual behavior**
The PNG loader totally fails, killing the pipeline. 

**Logs**
Here's an example where I used a decorated stream to back-fill the PNG.
Stream-1 is a Buffered Stream of sorts, which keeps a nearly 8k slice buffered.
Stream-2 is a the actual Source stream, and its direct calls.

```
Stream-1:-> 8096 bytes
Stream-1:<- 8096 bytes
	Stream-2:-> 8 bytes
	Stream-2:<- 8 bytes
	Stream-2:-> 29 bytes
	Stream-2:<- 29 bytes
	Stream-2:-> 4 bytes
	Stream-2:<- 4 bytes
	Stream-2:-> 273 bytes
	Stream-2:<- 273 bytes
	Stream-2:-> 4 bytes
	Stream-2:<- 4 bytes
	Stream-2:-> 8 bytes
	Stream-2:<- 8 bytes
	Stream-2:-> 8192 bytes
	Stream-2:<- 7770 bytes
errors:
pngload_source: end of stream
vips2png: unable to write to target target_custom
```

Stream-2 in this case now uses a backfill approach. If a PNG is being read and less data is returned than requested, additional requests are made.

```
Stream-1:-> 8096 bytes
Stream-1:<- 8096 bytes
	Stream-2:-> 8 bytes
	Stream-2:<- 8 bytes
	Stream-2:-> 29 bytes
	Stream-2:<- 29 bytes
	Stream-2:-> 4 bytes
	Stream-2:<- 4 bytes
	Stream-2:-> 273 bytes
	Stream-2:<- 273 bytes
	Stream-2:-> 4 bytes
	Stream-2:<- 4 bytes
	Stream-2:-> 8 bytes
	Stream-2:<- 8 bytes
	Stream-2:-> 8192 bytes
	Stream-2:<- 7770 bytes
	Stream-2: <~> reading an additional 422 bytes from 7770/8192
	Stream-2:-> 422 bytes
Stream-1:-> 422 bytes
Stream-1:<- 422 bytes
	Stream-2:<- 422 bytes
	Stream-2:-> 8192 bytes
Stream-1:-> 8192 bytes
Stream-1:<- 8192 bytes
	Stream-2:<- 8192 bytes
	Stream-2:-> 3293 bytes
Stream-1:-> 3293 bytes
Stream-1:<- 3293 bytes
	Stream-2:<- 3293 bytes
```

**Environment**
Confirmed on Linux and macOS.
Both were using version 8.10.1 from [libvips-packaging](https://github.com/kleisauke/libvips-packaging). A similar issue exists with older libvips version as well when using libpng.

**Example code**
I Was able to write some shareable C# to demonstrate this issue quickly
```
`23
Executing sample: Network stream
-> 8 bytes
<- 8 bytes
-> 29 bytes
<- 29 bytes
-> 4 bytes
<- 4 bytes
<NetVips.Image 800x600 uchar, 4 bands, srgb>
-> 8192 bytes
<- 3173 bytes
Unhandled exception. NetVips.VipsException: unable to call VipsForeignSaveJpegTarget
pngload_source: end of stream

   at NetVips.Operation.Call(String operationName, VOption kwargs, Image matchImage, Object[] args) in /Users/vibbix/git/net-vips/src/NetVips/Operation.cs:line 229
   at NetVips.ExtensionMethods.Call(Image image, String operationName, VOption kwargs, Object[] args) in /Users/vibbix/git/net-vips/src/NetVips/ExtensionMethods.cs:line 92
   at NetVips.Image.WriteToTarget(Target target, String formatString, VOption kwargs) in /Users/vibbix/git/net-vips/src/NetVips/Image.cs:line 1008
   at NetVips.Image.WriteToStream(Stream stream, String formatString, VOption kwargs) in /Users/vibbix/git/net-vips/src/NetVips/Image.cs:line 1027
   at NetVips.Samples.NetworkStream.Execute(String[] args) in /Users/vibbix/git/net-vips/samples/NetVips.Samples/Samples/NetworkStream.cs:line 27
   at NetVips.Program.Main(String[] args) in /Users/vibbix/git/net-vips/samples/NetVips.Samples/Program.cs:line 52

Process finished with exit code 6.
`
```
[Demonstrates_how_to_replicate_the_png_loader_issue.txt](https://github.com/libvips/libvips/files/5564047/Demonstrates_how_to_replicate_the_png_loader_issue.txt)

**Comment by jcupitt** [comment]
Hello @vibbix,

Thanks for the detailed report. I think this was fixed a week or so ago with https://github.com/libvips/libvips/commit/0a91e2d8c7a3c1062e7d94205dc924ccd075c459, see https://github.com/kleisauke/net-vips/issues/101

The fix is in the head of the 8.10 branch and should be released in 8.10.3 in a few day's time.

**Comment by kleisauke** [comment]
[NetVips.Native v8.10.5(.1)](https://www.nuget.org/packages/NetVips.Native/8.10.5.1) is now available. The network stream sample was extended with commit https://github.com/kleisauke/net-vips/commit/2b8d25d330373a1dee44d346121671db12181d8b to test this. Let me know if this works for you!
